File: frontend/src/lib/api/quiz.ts
import { apiClient } from "./client";

export interface QuizSet {
  id: string;
  owner_id: string;
  document_id?: string;
  document_name?: string | null;
  selected_document_ids?: string[] | null;
  title?: string | null;
  question_count: number;
  created_at: string;
  updated_at: string;
  best_score?: number | null;
  last_score?: number | null;
  attempt_count?: number;
  is_favorite?: boolean;
}

export interface QuizQuestion {
  id: string;
  quiz_set_id: string;
  question: string;
  options: string[];
  correct_answer: number;
  explanation?: string;
  topic?: string;
}

export const quizApi = {
  listSets: () => apiClient<QuizSet[]>("/quiz/sets", {
    method: "GET"
  }),

  getSet: (id: string) => apiClient<QuizSet & { questions: QuizQuestion[] }>(`/quiz/sets/${id}`, {
    method: "GET"
  }),

  deleteSet: (id: string) => apiClient<void>(`/quiz/sets/${id}`, {
    method: "DELETE"
  }),

  updateSet: (id: string, body: { title?: string; is_favorite?: boolean }) => apiClient<QuizSet>(`/quiz/sets/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  }),

  generate: (body: { document_id?: string; document_ids?: string[] | null; title?: string; num_questions?: number }) =>
    apiClient<{ quiz_set_id: string; questions: any[]; sources: any[] }>(`/quiz/generate`, {
      method: "POST",
      body: JSON.stringify(body),
    }),

  submitAttempt: (setId: string, score: number, percentage: number) =>
    apiClient(`/quiz/sets/${setId}/attempts`, {
      method: "POST",
      body: JSON.stringify({ score, percentage }),
    }),

  createAdaptive: (setId: string, questionIds: string[]) =>
    apiClient<QuizSet & { questions: QuizQuestion[] }>(`/quiz/sets/${setId}/adaptive`, {
      method: "POST",
      body: JSON.stringify({ question_ids: questionIds }),
    }),

  getAnalytics: () => apiClient<QuizAnalyticsResponse>("/quiz/analytics", {
    method: "GET"
  }),
};

export interface QuizAnalyticsResponse {
  generated: number;
  attempted: number;
  average_score: number | null;
  best_score: number | null;
  completion_rate: number;
}

File: backend/app/schemas/quiz.py
"""Quiz schemas."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class QuizQuestionResponse(BaseModel):
    """Quiz Question response."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    quiz_set_id: UUID
    question: str
    options: list[str]
    correct_answer: int
    explanation: str | None = None
    topic: str | None = None


class QuizSetResponse(BaseModel):
    """Quiz Set response."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    owner_id: UUID
    document_id: UUID | None = None
    document_name: str | None = None
    selected_document_ids: list[str] | None = None
    title: str | None = None
    question_count: int
    created_at: datetime
    updated_at: datetime
    best_score: int | None = None
    last_score: int | None = None
    attempt_count: int = 0
    is_favorite: bool = False


class QuizAttemptCreate(BaseModel):
    """Request body for creating a quiz attempt."""

    score: int
    percentage: int


class QuizAttemptResponse(BaseModel):
    """Response for a quiz attempt."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    quiz_set_id: UUID
    score: int
    percentage: int
    created_at: datetime


class QuizSetDetailResponse(QuizSetResponse):
    """Quiz Set response with questions."""

    questions: list[QuizQuestionResponse]


class QuizSetUpdateRequest(BaseModel):
    """Request body for updating a Quiz Set."""

    title: str | None = None
    is_favorite: bool | None = None


class AdaptiveQuizRequest(BaseModel):
    """Request body for creating an adaptive quiz."""

    question_ids: list[UUID]

class QuizAnalyticsResponse(BaseModel):
    """Quiz Analytics response."""
    
    generated: int
    attempted: int
    average_score: float | None = None
    best_score: int | None = None
    completion_rate: float
